### PR TITLE
Add and use generic functions  in order to gather cmake code

### PR DESCRIPTION
**Pull request by gregoire-dl**
## Features list

- [x] Add and use generic functions in order to gather cmake code
- [x] Use `GNUInstallDirs` install directory variables
- [x] Add software version
- [x] Generate Windows versioning information (.rc)

### CODE REVIEW

File: src/cmake/Helpers.cmake [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  target_include_directories(${TEST_EXECUTABLE_NAME}
+    PUBLIC ${TEST_INCLUDE_DIRS}
+           ${Boost_INCLUDE_DIRS}
+  )
+
+  set_property(TARGET ${TEST_EXECUTABLE_NAME}
+    PROPERTY FOLDER Test
+  )
+
+  add_test(NAME ${TEST_EXECUTABLE_NAME}
```
Comment (fabiencastan) [inline src/cmake/Helpers.cmake]: add_test should not have the same name as the add_executable:
```
add_test(NAME "aliceVision_testexec_${TEST_NAME}"
         WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
         COMMAND $<TARGET_FILE:${TEST_EXECUTABLE_NAME}> --catch_system_error=yes --log_level=all
)
```

File: src/cmake/Helpers.cmake [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+           WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
+           COMMAND $<TARGET_FILE:${TEST_EXECUTABLE_NAME}> --catch_system_error=yes --log_level=all
+  )
```
Comment (fabiencastan) [inline src/cmake/Helpers.cmake]: Would be good to declare the dependency between the test execution and the executable with `add_dependencies(<target> [<target-dependency>]...)`:
`add_dependencies("aliceVision_testexec_${TEST_NAME}" ${TEST_EXECUTABLE_NAME})`

File: CMakeLists.txt [hunk 3]
```diff
@@ -25,11 +25,18 @@ set(ALICEVISION_BUNDLE_PREFIX "${CMAKE_INSTALL_PREFIX}/bundle" CACHE STRING "Pat
 
 set(ALICEVISION_ROOT ${PROJECT_BINARY_DIR})
 
+# ==============================================================================
+# GNUInstallDirs CMake module
+# - Define GNU standard installation directories
+# - Provides install directory variables as defined by the GNU Coding Standards.
+# ==============================================================================
+include(GNUInstallDirs)
+
```
Comment (fabiencastan) [inline CMakeLists.txt]: Would be nice to add an option to disable it:
`option(ALICEVISION_USE_RPATH "Add RPATH on softwares with relative paths to libraries" ON)`

```
if(ALICEVISION_USE_RPATH)
...
endif()
```

File: src/cmake/Helpers.cmake (hunk 2 above)
Comment (gregoire-dl) [reply src/cmake/Helpers.cmake]: Sorry but I don't understand.
I can't do like your example. In `add_test` NAME is not a target.
```
  CMake Error at src/cmake/Helpers.cmake:193 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "aliceVision_testexec_foo".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
```
What did I misunderstand ?

File: CMakeLists.txt (hunk 3 above)
Comment (simogasp) [reply CMakeLists.txt]: `software`, it is uncountable

File: src/aliceVision/fuseCut/CMakeLists.txt [hunk 4]
```diff
@@ -46,15 +35,8 @@ target_link_libraries(aliceVision_fuseCut
...
-
-set_property(TARGET aliceVision_fuseCut
-  PROPERTY FOLDER AliceVision
-)
-
-install(TARGETS aliceVision_fuseCut
-  DESTINATION lib
-  EXPORT aliceVision-targets
+  PRIVATE_INCLUDE_DIRS
+    $<BUILD_INTERFACE:${NANOFLANN_INCLUDE_DIR}>
```
[the remark this answers is not in the document]
Comment (simogasp) [reply src/aliceVision/fuseCut/CMakeLists.txt]: yes if `nanoflann` is the imported target from find_package

File: src/cmake/version.rc.in [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#if @ALICEVISION_INSTALL_LIBRARY@
+            VALUE "FileDescription",  "@ALICEVISION_INSTALL_NAME@ - AliceVision with @CMAKE_CXX_COMPILER_ID@ @CMAKE_CXX_COMPILER_VERSION@\0"
+#else // software
+            VALUE "FileDescription",  "@ALICEVISION_INSTALL_NAME@ - AliceVision @ALICEVISION_VERSION_MAJOR@.@ALICEVISION_VERSION_MINOR@.@ALICEVISION_VERSION_REVISION@\0"
+#endif
+            VALUE "FileVersion",      "@ALICEVISION_INSTALL_VERSION_MAJOR@.@ALICEVISION_INSTALL_VERSION_MINOR@.@ALICEVISION_INSTALL_VERSION_REVISION@.0\0"
+            VALUE "InternalName",     "@ALICEVISION_INSTALL_NAME@\0"
+            VALUE "LegalCopyright",   "Copyright (c) 2018 AliceVision contributors\0"
+            VALUE "ProductName",      "AliceVision Library | @ALICEVISION_INSTALL_NAME@\0"
+            VALUE "ProductVersion",   "@ALICEVISION_INSTALL_VERSION_MAJOR@.@ALICEVISION_INSTALL_VERSION_MINOR@.@ALICEVISION_INSTALL_VERSION_REVISION@.0\0"
```
Comment (zvrba) [inline src/cmake/version.rc.in]: It is superfluous to duplicate the same version string in both FileVersion and ProductVersion. It'd be better to use ProductVersion to store git commit number instead.

**Pull request merged**